pulser-simulation/pulser_simulation/simulation.py

```diff
@@ -201,14 +201,25 @@ def __init__(
                 self._meas_basis = self.basis_name.replace("_with_error", "")
         self.set_initial_state("all-ground")
 
-    @cached_property
-    def _noiseless_hamiltonian(self) -> Hamiltonian:
+    @lru_cache(maxsize=2)
+    def _noiseless_hamiltonian(self, leakage: bool) -> Hamiltonian:
```

Comment:
To really minimize disruption (and potential confusion, since leakage being included in something named "noiseless" is counterintuitive), how about we do this?

```suggestion
    @property
    def _noiseless_hamiltonian(self, leakage) -> Hamiltonian:
        return self._get_noiseless_hamiltonian(leakage=False)
        
    @lru_cache(maxsize=2)
    def _get_noiseless_hamiltonian(self, leakage: bool) -> Hamiltonian:
```

You can then use this in `QutipBackendV2`

```
self._get_noiseless_hamiltonian(leakage=noise_model.with_leakage)
```